crates/containerd-shim-wasm/src/sandbox/containerd/client.rs

```diff
@@ -394,126 +392,140 @@ impl Client {
-                        let mut original_layer = self.get_info(original_config.digest())?;
-                        original_layer
-                            .labels
-                            .insert(precompile_id.clone(), precompiled_content.digest.clone());
-                        original_layer.labels.insert(
-                            format!("containerd.io/gc.ref.content.precompile.{}", i),
-                            precompiled_content.digest.clone(),
-                        );
-                        self.update_info(original_layer)?;
-
-                        // The original image is considered a root object, by adding a ref to the new compiled content
-                        // We tell containerd to not garbage collect the new content until this image is removed from the system
-                        // this ensures that we keep the content around after the lease is dropped
-                        // We also save the precompiled flag here since the image labels can be mutated containerd, for example if the image is pulled twice
-                        log::debug!(
-                            "updating image content with precompile digest to avoid garbage collection"
-                        );
-                        let mut image_content = self.get_info(&image_digest)?;
-                        image_content.labels.insert(
-                            format!("containerd.io/gc.ref.content.precompile.{}", i),
-                            precompiled_content.digest,
-                        );
-                        image_content
-                            .labels
-                            .insert(precompile_id.clone(), "true".to_string());
-                        self.update_info(image_content)?;
-
-                        layers_for_runtime.push(WasmLayer {
-                            config: original_config.clone(),
-                            layer: compiled_layer.clone(),
-                        });
-                    }
-                    return Ok((layers_for_runtime, platform));
+                    compiled_layers
                 }
                 Err(e) => {
                     log::error!("precompilation failed: {}", e);
+                    return Ok((layers, platform));
                 }
             };
-        }
 
-        if layers.is_empty() {
-            log::info!("no WASM layers found in OCI image");
-            return Ok((vec![], platform));
-        }
+            let mut layers_for_runtime = Vec::with_capacity(compiled_layers.len());
+            for (i, compiled_layer) in compiled_layers.iter().enumerate() {
+                if compiled_layer.is_none() {
+                    log::debug!("no compiled layer using original");
+                    layers_for_runtime.push(layers[i].clone());
+                    continue;
+                }
+
+                let compiled_layer = compiled_layer.as_ref().unwrap();
+                let original_config = &compiled_layer.config;
```

Comment:
Seems weird to consider the config returned from `precompile` be "original". This is why i think keeping the API to `Option<Vec<u8>>` may make sense. Otherwise, this may want to use the `compiled_layers` above or add a comment to the API saying that it expects the returned config to be the original one.

Reply:
Reverted back